src/Voron/Data/Graphs/Hnsw.cs

```diff
@@ -0,0 +1,1315 @@
+                    case Constants.Graphs.VectorId.PostingList:
+                        return UpdatePostingList(rawPostingListId, in modifications, pforEncoder, ref pforDecoder, ref listBuffer);
+                }
+                
+                PostingList.SortEntriesAndRemoveDuplicatesAndRemovals(ref listBuffer);
+                if (listBuffer.Count is 0 or 1)
+                {
+                    if (hasSmallPostingList)
+                    {
+                        Container.Delete(_searchState.Llt, _searchState.Options.Container, rawPostingListId);
+                    }
+
+                    if (listBuffer.Count is 0) 
+                        return 0;
+                    
+                    Debug.Assert((listBuffer[0] & Constants.Graphs.VectorId.PostingList) == 0, "(listBuffer[0] & 0b11) == 0");
+                    return listBuffer[0] | Constants.Graphs.VectorId.Single;
+                }
+
+                int size = pforEncoder.Encode(listBuffer.RawItems, listBuffer.Count);
+                if (size > Container.MaxSizeInsideContainerPage)
+                {
+                    DeleteOldSmallPostingListIfNeeded();
+                    return CreateNewPostingList(pforEncoder);
+                }
+
+                byteBuffer.EnsureCapacityFor(size + 5);
+                var offset = VariableSizeEncoding.Write(byteBuffer.RawItems, listBuffer.Count);
+                (int itemsCount, int sizeUsed) = pforEncoder.Write(byteBuffer.RawItems + offset, byteBuffer.Capacity - offset);
+                byteBuffer.Count = sizeUsed + offset;
+                Debug.Assert(itemsCount == listBuffer.Count && sizeUsed == size, "itemsCount == listBuffer.Count && sizeUsed == size");
+                Span<byte> mutable;
+                if (currentSize == byteBuffer.Count)
+                {
+                    mutable = Container.GetMutable(_searchState.Llt, rawPostingListId);
+                }
+                else
+                {
+                    DeleteOldSmallPostingListIfNeeded();
+                    rawPostingListId = Container.Allocate(_searchState.Llt, _searchState.Options.Container, byteBuffer.Count, out mutable);
+                }
+                Span<byte> span = byteBuffer.ToSpan();
+                span.CopyTo(mutable);
+
+                Debug.Assert((rawPostingListId & Constants.Graphs.VectorId.PostingList) == 0, "(rawPostingListId & 0b11) == 0");
+                return rawPostingListId | Constants.Graphs.VectorId.SmallPostingList;
+                
+                
+                void DeleteOldSmallPostingListIfNeeded()
+                {
+                    if (hasSmallPostingList)
+                    {
+                        Container.Delete(_searchState.Llt, _searchState.Options.Container, rawPostingListId);
+                    }
+                }
+            }
+        }
+        
+        public void Dispose()
+        {
+            //todo: we may wants to release the vector hash cache
```

Comment:
TODO

Reply:
@redknightlois 
Can I safely release the memory of the ByteString keys from dictionary?
eg.
```csharp
            foreach (var key in _vectorHashCache.Keys)
            {
                var k = key;
                _searchState.Llt.Allocator.Release(ref k);
            }
            
            _vectorHashCache.Clear();
```

Reply:
You can release the memory of the `ByteString` keys *if and only if* you adhere to these principles:

Lifecycle Ownership: The moment you put something into the dictionary, the dictionary becomes its owner. This means you relinquish all other references. If you kept a reference somewhere else, you’re setting yourself up for undefined behavior—so don’t.

Transaction Context: The release must happen in the same transaction where the keys were allocated. Cross-transaction operations with memory management? That’s a recipe for disaster.

Iterator Behavior: Iterators need to play nice. If the iterator is going to touch or dereference keys they must not have been disposed, and the one that iterates must not keep those pointer.